llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp

```diff
@@ -3745,6 +3745,21 @@ AMDGPURegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
     if (!DstBank)
       DstBank = SrcBank;
 
+    // The calling convention is to be updated such that i1 function arguments
+    // or return values are assigned to SGPRs without promoting to i32. With
+    // this, for i1 function arguments, the call of getRegBank() above gives
+    // incorrect result. We set both src and dst banks to VCCRegBank.
+    if (!MI.getOperand(1).getReg().isVirtual() &&
```

Comment:
Shouldn't directly depend on isVirtual / isPhysical. You should be relying on the bank retrieved above which should be correct

Reply:
Actually, once the calling convention for i1 arg is changed, we would have instructions such as `%0:_(s1) = COPY $sgpr4_sgpr5`. For this instruction, after the 2 calls of `getRegBank()`, `DstBank` is null, and `SrcBank` is `SGPRRegBank`. We want both to be `VCCRegBank`, hence the patch.

The reason it returns `SGPRRegBank` instead of `VCCRegBank` is because of a flaw in `RegisterBankInfo::getRegBank()`, as follows:
```C
const RegisterBank *
RegisterBankInfo::getRegBank(Register Reg, const MachineRegisterInfo &MRI,
                             const TargetRegisterInfo &TRI) const {
  if (!Reg.isVirtual()) {
    // FIXME: This was probably a copy to a virtual register that does have a
    // type we could use.
    const TargetRegisterClass *RC = getMinimalPhysRegClass(Reg, TRI);
    return RC ? &getRegBankFromRegClass(*RC, LLT()) : nullptr;  
  }
...
```
Note that there's a FIXME. In the call of `getRegBankFromRegClass` it passes the default type `LLT()`, which is not what we want here.
The IF statement in the patch basically says, if the src is physical and dst type is s1, then set the reg bank to VCCRegBank, which corrects the problem.

Reply:
One possibly missing piece in the IF is `MI.isCopy()` because the code actually can be reached for both COPY and FREEZE.

Reply:
If you really think I should get rid of the call of `isVirtual()`, I've tried the following. All tests have passed, but the condition is less limiting than the original version:
```diff
@@ -3735,32 +3735,30 @@ AMDGPURegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
   const MachineRegisterInfo &MRI = MF.getRegInfo();

   if (MI.isCopy() || MI.getOpcode() == AMDGPU::G_FREEZE) {
     // The default logic bothers to analyze impossible alternative mappings. We
     // want the most straightforward mapping, so just directly handle this.
     const RegisterBank *DstBank = getRegBank(MI.getOperand(0).getReg(), MRI,
                                              *TRI);
     const RegisterBank *SrcBank = getRegBank(MI.getOperand(1).getReg(), MRI,
                                              *TRI);
     assert(SrcBank && "src bank should have been assigned already");
-    if (!DstBank)
-      DstBank = SrcBank;

-    // The calling convention is to be updated such that i1 function arguments
-    // or return values are assigned to SGPRs without promoting to i32. With
-    // this, for i1 function arguments, the call of getRegBank() above gives
-    // incorrect result. We set both src and dst banks to VCCRegBank.
-    if (!MI.getOperand(1).getReg().isVirtual() &&
-        MRI.getType(MI.getOperand(0).getReg()) == LLT::scalar(1)) {
+    // For copy from physical reg to s1 dest, the call of getRegBank() above
+    // gives incorrect result. We set both src and dst banks to VCCRegBank.
+    if (MI.isCopy() && !DstBank && MRI.getType(MI.getOperand(0).getReg()) == LLT::scalar(1)) {
       DstBank = SrcBank = &AMDGPU::VCCRegBank;
     }

+    if (!DstBank)
+      DstBank = SrcBank;
+
```
Pls let me know your thoughts.

Reply:
You need to add MIR tests showing the regclass / regbank / physreg with i1 permutations here first to review this usefully. You shouldn't need to check isCopy again either. Freeze is just the same as copy, except it can't have the physreg inputs